FE Restart Failed with exception: com.sleepycat.je.rep.InsufficientLogException

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
Not easy to reproduce this bug.

### Expected behavior (Required)
restart successfully.

### Real behavior (Required)
restart failed.
```java
2022-04-28 16:38:15,491 ERROR (UNKNOWN 172.26.92.139_29917_1650957989140(-1)|1) [StarRocksFE.start():130] StarRocksFE start failed
com.sleepycat.je.rep.InsufficientLogException: (JE 7.3.8) Environment must be closed, caused by: com.sleepycat.je.rep.InsufficientLogException: Environment invalid because of previous exception: (JE 7.3.8) 172.26.92.139_29917_1650957989140(2407):/home/disk4/gengjun/fe/doris-meta/bdb INSUFFICIENT_LOG: Log files at this node are obsolete. Environment is invalid and must be closed.refreshVLSN=null logProviders=null repImpl=null props=null
        at com.sleepycat.je.rep.InsufficientLogException.wrapSelf(InsufficientLogException.java:315) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.log.LogManager.getLogEntry(LogManager.java:830) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.log.LogManager.getLogEntry(LogManager.java:784) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.log.LogManager.getLogEntryHandleFileNotFound(LogManager.java:935) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.DiskOrderedScanner.fetchEntry(DiskOrderedScanner.java:2062) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.DiskOrderedScanner.fetchAndProcessBINs(DiskOrderedScanner.java:1634) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.DiskOrderedScanner.scanSerial(DiskOrderedScanner.java:783) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.DiskOrderedScanner.scan(DiskOrderedScanner.java:703) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.dbi.DatabaseImpl.count(DatabaseImpl.java:2266) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.Database.count(Database.java:1910) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.starrocks.journal.bdbje.BDBJEJournal.getMaxJournalId(BDBJEJournal.java:223) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJEJournal.open(BDBJEJournal.java:299) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.open(EditLog.java:829) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.Catalog.initialize(Catalog.java:884) ~[starrocks-fe.jar:?]
        at com.starrocks.StarRocksFE.start(StarRocksFE.java:107) [starrocks-fe.jar:?]
        at com.starrocks.StarRocksFE.main(StarRocksFE.java:62) [starrocks-fe.jar:?]
Caused by: com.sleepycat.je.rep.InsufficientLogException: Environment invalid because of previous exception: (JE 7.3.8) 172.26.92.139_29917_1650957989140(2407):/home/disk4/gengjun/fe/doris-meta/bdb INSUFFICIENT_LOG: Log files at this node are obsolete. Environment is invalid and must be closed. Originally thrown by HA thread: REPLICA 172.26.92.139_29917_1650957989140(2407)
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.setupLogRefresh(ReplicaFeederSyncup.java:664) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.getFeederRecord(ReplicaFeederSyncup.java:732) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.findMatchpoint(ReplicaFeederSyncup.java:416) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.execute(ReplicaFeederSyncup.java:151) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.impl.node.Replica.initReplicaLoop(Replica.java:711) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoopInternal(Replica.java:474) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoop(Replica.java:409) ~[starrocks-bdb-je-7.3.8.jar:?]
        at com.sleepycat.je.rep.impl.node.RepNode.run(RepNode.java:1873) ~[starrocks-bdb-je-7.3.8.jar:?]
```

### StarRocks version (Required)
- all

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -241,6 +241,12 @@ public class BDBEnvironment {
         restore.execute(insufficientLogEx, config);
     }
 
+    public void refreshAndSetup(InsufficientLogException insufficientLogEx) {
+        refreshLog(insufficientLogEx);
+        close();
+        setup();
+    }
+
     public ReplicationGroupAdmin getReplicationGroupAdmin() {
         return this.replicationGroupAdmin;
     }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -309,6 +309,13 @@ public class BDBJEJournal implements Journal {
                 // Copy the missing log files from a member of the replication group who owns the files
                 LOG.warn("catch insufficient log exception. will recover and try again.", insufficientLogEx);
                 bdbEnvironment.refreshAndSetup(insufficientLogEx);
+            } catch (Throwable t) {
+                LOG.warn("catch exception, sleep and retry", t);
+                try {
+                    Thread.sleep(5000L);
+                } catch (InterruptedException ie) {
+
+                }
             }
         }
     }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -27,8 +27,6 @@ import com.sleepycat.je.DatabaseEntry;
 import com.sleepycat.je.DatabaseException;
 import com.sleepycat.je.OperationStatus;
 import com.sleepycat.je.rep.InsufficientLogException;
-import com.sleepycat.je.rep.NetworkRestore;
-import com.sleepycat.je.rep.NetworkRestoreConfig;
 import com.starrocks.common.Pair;
 import com.starrocks.common.io.DataOutputBuffer;
 import com.starrocks.common.io.Writable;

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -280,6 +280,11 @@ public class BDBJEJournal implements Journal {
         List<Long> dbNames = null;
         for (int i = 0; i < RETRY_TIME; i++) {
             try {
+                // sleep for retry
+                if (i > 0) {
+                    Thread.sleep(3000L);
+                }
+
                 dbNames = bdbEnvironment.getDatabaseNames();
 
                 if (dbNames == null) {
@@ -310,12 +315,7 @@ public class BDBJEJournal implements Journal {
                 LOG.warn("catch insufficient log exception. will recover and try again.", insufficientLogEx);
                 bdbEnvironment.refreshAndSetup(insufficientLogEx);
             } catch (Throwable t) {
-                LOG.warn("catch exception, sleep and retry", t);
-                try {
-                    Thread.sleep(5000L);
-                } catch (InterruptedException ie) {
-
-                }
+                LOG.warn("catch exception, retried: {} ", i, t);
             }
         }
     }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -26,6 +26,9 @@ import com.sleepycat.je.Database;
 import com.sleepycat.je.DatabaseEntry;
 import com.sleepycat.je.DatabaseException;
 import com.sleepycat.je.OperationStatus;
+import com.sleepycat.je.rep.InsufficientLogException;
+import com.sleepycat.je.rep.NetworkRestore;
+import com.sleepycat.je.rep.NetworkRestoreConfig;
 import com.starrocks.common.Pair;
 import com.starrocks.common.io.DataOutputBuffer;
 import com.starrocks.common.io.Writable;
@@ -275,28 +278,41 @@ public class BDBJEJournal implements Journal {
         }
 
         // Open a new journal database or get last existing one as current journal database
-        List<Long> dbNames = bdbEnvironment.getDatabaseNames();
-        if (dbNames == null) {
-            LOG.error("fail to get dbNames while open bdbje journal. will exit");
-            System.exit(-1);
-        }
-        if (dbNames.size() == 0) {
-            /*
-             *  This is the very first time to open. Usually, we will open a new database named "1".
-             *  But when we start cluster with an image file copied from other cluster,
-             *  here we should open database with name image max journal id + 1.
-             *  (default GlobalStateMgr.getCurrentState().getReplayedJournalId() is 0)
-             */
-            String dbName = Long.toString(GlobalStateMgr.getCurrentState().getReplayedJournalId() + 1);
-            LOG.info("the very first time to open bdb, dbname is {}", dbName);
-            currentJournalDB = bdbEnvironment.openDatabase(dbName);
-        } else {
-            // get last database as current journal database
-            currentJournalDB = bdbEnvironment.openDatabase(dbNames.get(dbNames.size() - 1).toString());
-        }
+        Pair<String, Integer> helperNode = GlobalStateMgr.getCurrentState().getHelperNode();
+        List<Long> dbNames = null;
+        for (int i = 0; i < RETRY_TIME; i++) {
+            try {
+                dbNames = bdbEnvironment.getDatabaseNames();
 
-        // set next journal id
-        nextJournalId.set(getMaxJournalId() + 1);
+                if (dbNames == null) {
+                    LOG.error("fail to get dbNames while open bdbje journal. will exit");
+                    System.exit(-1);
+                }
+                if (dbNames.size() == 0) {
+                    /*
+                     *  This is the very first time to open. Usually, we will open a new database named "1".
+                     *  But when we start cluster with an image file copied from other cluster,
+                     *  here we should open database with name image max journal id + 1.
+                     *  (default Catalog.getCurrentCatalog().getReplayedJournalId() is 0)
+                     */
+                    String dbName = Long.toString(GlobalStateMgr.getCurrentState().getReplayedJournalId() + 1);
+                    LOG.info("the very first time to open bdb, dbname is {}", dbName);
+                    currentJournalDB = bdbEnvironment.openDatabase(dbName);
+                } else {
+                    // get last database as current journal database
+                    currentJournalDB = bdbEnvironment.openDatabase(dbNames.get(dbNames.size() - 1).toString());
+                }
+
+                // set next journal id
+                nextJournalId.set(getMaxJournalId() + 1);
+
+                break;
+            } catch (InsufficientLogException insufficientLogEx) {
+                // Copy the missing log files from a member of the replication group who owns the files
+                LOG.warn("catch insufficient log exception. will recover and try again.", insufficientLogEx);
+                bdbEnvironment.refreshAndSetup(insufficientLogEx);
+            }
+        }
     }
 
     @Override

